### PR TITLE
ocamlPackages.eio: init at 0.10; mtime: 1.4.0 → 2.0.0; iomux init at 0.3; domain-local-await: init at 0.2.0

### PR DESCRIPTION
**Pull request by toastal**
###### Description of changes

`eio`: https://github.com/ocaml-multicore/eio

required upgrade to `mtime` (separate commit): https://github.com/dbuenzli/mtime
required `iomux` as a dependency for `eio_posix` (separate commit): https://github.com/haesbaert/ocaml-iomux

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by anmonteiro — CHANGES_REQUESTED** [review]
mtime 2.0 breaks irmin packages, index, progress

**Review by anmonteiro — CHANGES_REQUESTED** [review]
mtime 2.0 breaks irmin packages, index, progress

**Review by anmonteiro — CHANGES_REQUESTED** [review]
mtime 2.0 breaks irmin packages, index, progress

**Comment by toastal** [comment]
@vbgl @Alizter y’all have been very helpful so far with reviewing OCaml-related stuff. This compiled on my x86_64 machine (AMD Ryzen Zen3). I’m usually a bit confused by which `*Inputs` are which.

File: pkgs/development/ocaml-modules/eio/default.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    bigstringaf
+    cstruct
+    fmt
+    hmap
+    lwt-dllist
+    mtime
+    optint
+    psq
+  ] ++ lib.optionals stdenv.isLinux [
+    uring
```
Comment (anmonteiro) [inline pkgs/development/ocaml-modules/eio/default.nix:42]: eio doesn't depend on uring, eio_linux does.

Comment (toastal) [reply pkgs/development/ocaml-modules/eio/default.nix:42]: Oh, do I need to create a series of separate packages then?

**Comment by toastal** [comment]
>  mtime 2.0 breaks irmin packages, index, progress

@anmonteiro What does this mean &hellip; you lost me at “index, progress”. Does this mean it’s being worked on?

File: pkgs/development/ocaml-modules/eio/default.nix (hunk 1 above)
Comment (toastal) [reply pkgs/development/ocaml-modules/eio/default.nix:42]: Removed.

**Comment by anmonteiro** [comment]
> > mtime 2.0 breaks irmin packages, index, progress
> 
> @anmonteiro What does this mean … you lost me at “index, progress”. Does this mean it’s being worked on?

`ocamlPackages.{irmin,irmin-containers,irmin-pack,progress,index}`

**Comment by anmonteiro** [comment]
> Does this mean it’s being worked on?

One potential solution is adding `mtime_1` like I did in https://github.com/nix-ocaml/nix-overlays/pull/590

File: pkgs/development/ocaml-modules/eio/default.nix (hunk 1 above)
Comment (anmonteiro) [reply pkgs/development/ocaml-modules/eio/default.nix:42]: right. check https://github.com/nix-ocaml/nix-overlays/tree/master/ocaml/eio for inspiration.

**Comment by toastal** [comment]
@anmonteiro I added a commit in the middle for 1.x. Does it suffice?

**Comment by toastal** [comment]
(My goal of getting eio in is to use it to learn OCaml @ v5.x. With multicore support finally here, it’s a good time to learn, but I really don’t know the language or ecosystem (yet)).

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>acgtk</li>
    <li>ligo</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>comby</li>
    <li>dune-release</li>
    <li>jackline</li>
    <li>liquidsoap</li>
  </ul>
</details>

**Comment by ulrikstrid** [comment]
`ligo` in my comment above actually is i `irmin-test` that is getting both `mtime@2` and `mtime@1.x`

**Comment by ulrikstrid** [comment]
If you want to I can try to figure out the `irmin`-stuff, I understand that it can be daunting trying to fix a ecosystem that you're not familiar with (yet).

**Comment by toastal** [comment]
@ulrikstrid any help would be appreciated. I just want to use the latest tool to learn 😅

**Comment by ulrikstrid** [comment]
I took the liberty of pushing changes needed for ligo and acgtk, feel free to squash them into the appropriate commit

**Comment by ulrikstrid** [comment]
Hmm, there are more failures after fixing these. I wonder if it's better to do it the other way around and introduce `mtime_2` and use that where possible

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>27 packages failed to build:</summary>
  <ul>
    <li>comby</li>
    <li>jackline</li>
    <li>ocamlPackages.awa</li>
    <li>ocamlPackages.awa-lwt</li>
    <li>ocamlPackages.awa-mirage</li>
    <li>ocamlPackages.cohttp-mirage</li>
    <li>ocamlPackages.conduit-mirage</li>
    <li>ocamlPackages.dns-cli</li>
    <li>ocamlPackages.dns-client</li>
    <li>ocamlPackages.dns-client-lwt</li>
    <li>ocamlPackages.dns-client-mirage</li>
    <li>ocamlPackages.dns-stub</li>
    <li>ocamlPackages.git-mirage</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.happy-eyeballs-lwt</li>
    <li>ocamlPackages.happy-eyeballs-mirage</li>
    <li>ocamlPackages.http-mirage-client</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-mirage-git</li>
    <li>ocamlPackages.irmin-mirage-graphql</li>
    <li>ocamlPackages.letsencrypt-mirage</li>
    <li>ocamlPackages.mimic-happy-eyeballs</li>
    <li>ocamlPackages.mirage-crypto-rng-lwt</li>
    <li>ocamlPackages.plotkicadsch</li>
    <li>ocamlPackages.tls-lwt</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.iomux</li>
    <li>ocamlPackages.mtime</li>
    <li>ocamlPackages.mtime_1</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.ocsipersist</li>
    <li>ocamlPackages.ocsipersist-pgsql</li>
    <li>ocamlPackages.ocsipersist-sqlite</li>
    <li>ocamlPackages.opium</li>
  </ul>
</details>

**Comment by toastal** [comment]
Is it better to override `mtime_1` in the `top-level` file, or in the individual derivations? I see I did it one way but @ulrikstrid another way.

**Comment by ulrikstrid** [comment]
I used `mtime_1` directly in packages not in the `ocamlPackages` set and did the same thing you did when in the set.

I fixed some more failures now and rebased on master.

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>comby</li>
    <li>ocamlPackages.awa-lwt</li>
    <li>ocamlPackages.dns-cli</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.iomux</li>
    <li>ocamlPackages.mtime</li>
    <li>ocamlPackages.mtime_1</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.ocsipersist</li>
    <li>ocamlPackages.ocsipersist-pgsql</li>
    <li>ocamlPackages.ocsipersist-sqlite</li>
    <li>ocamlPackages.opium</li>
  </ul>
</details>

**Comment by ulrikstrid** [comment]
@vbgl could you look at this? It seems like everything builds now

File: pkgs/development/tools/comby/default.nix [hunk 2]
```diff
@@ -19,6 +19,10 @@ let
       minimalOcamlVersion = "4.08.1";
       doCheck = true;
 
+      # Allow conflicting mtime in findlib
+      # note(ulrikstrid): I can't find where mtime 2 comes from
```
Comment (vbgl) [inline pkgs/development/tools/comby/default.nix:23]: `opium`.

Comment (ulrikstrid) [reply pkgs/development/tools/comby/default.nix]: You are correct, thanks!

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>comby</li>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.iomux</li>
    <li>ocamlPackages.mtime</li>
    <li>ocamlPackages.mtime_1</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.ocsipersist</li>
    <li>ocamlPackages.ocsipersist-pgsql</li>
    <li>ocamlPackages.ocsipersist-sqlite</li>
    <li>ocamlPackages.opium</li>
  </ul>
</details>

**Comment by ulrikstrid** [comment]
@toastal I think we can squash some commits now to make it slightly cleaner.

Other than that I think it should be good to go, cc @vbgl

**Comment by ulrikstrid** [comment]
> @ulrikstrid Still unsure if `callPackage xxx { mtime = mtime_1 }` or using `mtime_1` inside the derivation is better. It seems this is doing a bit of both?

I answered that here I think: https://github.com/NixOS/nixpkgs/pull/230270#issuecomment-1554401363

**Comment by toastal** [comment]
I missed that. You think it's best to to squash all the `mtime`-related commits to one?

**Comment by vbgl** [comment]
Yes, you should squash them together.

**Comment by toastal** [comment]
This doesn’t build for me

```
error: builder for '/nix/store/kxymq9ji2vaii0r933f004a8pydpzlip-ocaml5.0.0-ocplib-endian-1.2.drv' failed with exit code 1;
       last 10 log lines:
       > File "src/dune", line 75, characters 26-31:
       > 75 |  (libraries ocplib_endian bytes))
       >                                ^^^^^
       > Error: Library "bytes" not found.
       > -> required by library "ocplib-endian.bigstring" in _build/default/src
       > -> required by _build/default/META.ocplib-endian
       > -> required by _build/install/default/lib/ocplib-endian/META
       > -> required by _build/default/ocplib-endian.install
       > -> required by alias install
       >
       For full logs, run 'nix log /nix/store/kxymq9ji2vaii0r933f004a8pydpzlip-ocaml5.0.0-ocplib-endian-1.2.drv'.
```

**Comment by toastal** [comment]
@ulrikstrid @vbgl @anmonteiro added the missing `domain-local-await` library. Also `0.10` was released (updated).

**Comment by ulrikstrid** [comment]
I think we just need a squash of the commits for a cleaner history and then we're good to go

**Comment by toastal** [comment]
@ulrikstrid done... I inputted my ssh key passphrase incorrectly XD

**Comment by ulrikstrid** [comment]
Normally it's `init at xxx`

**Comment by toastal** [comment]
So we use `->` (& `→`) to mean “to” but not `@` to mean “at”. Seems weird 🤷

**Comment by toastal** [comment]
commit message updated

**Comment by ulrikstrid** [comment]
> So we use `->` (& `→`) to mean “to” but not `@` to mean “at”. Seems weird 🤷

Yes, not sure why 😂

**Comment by ulrikstrid** [comment]
Going to do another pass on this a bit later today and if it's good I'll press merge

**Comment by toastal** [comment]
@ulrikstrid `domain-local-await` released 0.2.0 in the last day or so. Updated.

File: pkgs/development/ocaml-modules/eio/default.nix [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    mdx
+  ];
+
+  nativeCheckInputs = [
+    mdx.bin
+  ];
+
+  doCheck = true;
+
+  dontStrip = true;
```
Comment (ulrikstrid) [inline pkgs/development/ocaml-modules/eio/default.nix:55]: Why do we have this?

**Review by ulrikstrid — COMMENTED** [review]
Sorry for the delay, just a question about `dontStrip`, other than that I'm ready to merge this if it passes nixpkgs-review

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by ulrikstrid** [comment]
Result of `nixpkgs-review pr 230270` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/development/ocaml-modules/eio/default.nix (hunk 3 above)
Comment (toastal) [reply pkgs/development/ocaml-modules/eio/default.nix:55]: Copied from some other file 😅 When I read this weekend I was thinking the same thing. Can remove.

**Review by ulrikstrid — APPROVED** [review]

**Comment by ulrikstrid** [comment]
Updated the title to follow some other PRs. Will merge when CI is green.

**Comment by ulrikstrid** [comment]
I assume we don't want to backport this since we've renamed old mtime, any opinion about this @vbgl?

**Comment by anmonteiro** [comment]
Could be nice to add lwt_eio too: https://github.com/ocaml-multicore/lwt_eio

**Comment by anmonteiro** [comment]
`iomux` needs `hardeningDisable = [ "strictoverflow" ];` to work on macOS...

**Comment by toastal** [comment]
@anmonteiro I was looking at`lwt_eio` but this merge request had already ballooned enough as it was. I don’t run any Darwin-based devices so I have no idea what the requirements are.

**Pull request merged**